### ISSUE TITLE
Terraform remote-exec and file are using private IP addresses for connection

### ISSUE THREAD

**Issue by polasekr** [hashicorp/terraform#6394]
## Terraform Version

Terraform v0.6.15
## Affected Resource(s)
- aws_instance
## Terraform Configuration Files

```
/*******************************
* Kubernetes build server
********************************/
variable "env_file" {
    default = "k8s-bootstrap.env"
}
variable "instancefile" {
    default = ".instance.id"
}
variable "ssh_key" {
    default = "xxx-key.pem"
}

resource "aws_instance" "k8s-bootstrap" {
    ami = "${var.images.us-east-1}"
    instance_type = "t2.medium"
    subnet_id = "subnet-xxxxxx"
    vpc_security_group_ids = [ "sg-xxxxxxx","sg-yyyyyy" ]
    tags {
        Name = "k8s-bootstrap"
    }
    key_name = "deployment-key"
    provisioner "local-exec" {
        command = "echo export K8S_ENV_NAME=\\\"10point1\\\" AWS_S3_REGION=\\\"US Standard\\\" >${var.env_file}"
    }
    provisioner "local-exec" {
        command = "echo export KUBE_AWS_ZONE=\\\"${aws_subnet.k8s-172-31-1-0-s-subnet-1a.availability_zone}\\\" >>${var.env_file}"
    }
    provisioner "local-exec" {
        command = "echo export AWS_S3_BUCKET=\\\"kubernetes-artifacts-'$K8S_ENV_NAME'\\\" >>${var.env_file}"
    }
    provisioner "local-exec" {
        command = "echo export NUM_NODES=3 KUBE_OS_DISTRIBUTION=\\\"jessie\\\" >>${var.env_file}"
    }
    provisioner "local-exec" {
        command = "echo export INSTANCE_PREFIX=\\\"'$K8S_ENV_NAME'\\\" KUBE_AWS_INSTANCE_PREFIX=\\\"'$K8S_ENV_NAME'\\\" >>${var.env_file}"
    }
    provisioner "local-exec" {
        command = "echo export AWS_SSH_KEY='$HOME/.ssh/deployment_key_rsa' >>${var.env_file}"
    }
    provisioner "local-exec" {
        command = "echo export SERVICE_CLUSTER_IP_RANGE=\\\"10.0.0.0/16\\\" CLUSTER_IP_RANGE=\\\"10.244.0.0/16\\\" MASTER_IP_RANGE=\\\"10.246.0.0/24\\\" >>${var.env_file}"
    }
    provisioner "local-exec" {
        command = "echo export KUBE_SUBNET_CIDR=\\\"${aws_subnet.k8s-172-31-1-0-s-subnet-1a.cidr_block}\\\" >>${var.env_file}"
    }
    provisioner "remote-exec" {
        inline = [
           "echo ${aws_instance.k8s-bootstrap.id} >${var.instancefile}",
           "apt-get install awscli s3cmd",
           "s3cmd --access_key=${var.access_key} --secret_key=${var.secret_key} --ssl get s3://k8s-env/${aws_instance.k8s-bootstrap.id}",
           "mkdir .ssh",
           "chmod 600 .ssh",
           "s3cmd --access_key=${var.access_key} --secret_key=${var.secret_key} --ssl get s3://k8s-env/${var.ssh_key} .ssh/id_rsa",
           "s3cmd --access_key=${var.access_key} --secret_key=${var.secret_key} --ssl -r get s3://k8s-env/${var.account_id} ~/.aws",
           "git clone git@github.com:10point1/k8s_build_script.git",
           "cd k8s_build_script",
           "chmod 755 k8s_env.sh",
           "./k8s_env.sh ~/${aws_instance.k8s-bootstrap.id} up >k8s_installation.log 2>&1"
        ]
    }
}
resource "aws_eip" "k8s-bootstrap-IP" {
    //network_interface = "${aws_network_interface.k8s-bootstrap-net.id}"
    instance = "${aws_instance.k8s-bootstrap.id}"
    vpc = true
}
resource "aws_security_group" "ssh-in" {
  name = "ssh-in"
  description = "Allow http(s) and ssh traffic"
  vpc_id = "vpc-1ff9087a"
  ingress {
    from_port = 22
    to_port = 22
    protocol = "tcp"
    cidr_blocks = [ "0.0.0.0/0" ]
  }
  egress {
    from_port = 0
    to_port = 0
    protocol = "-1"
    cidr_blocks = [ "0.0.0.0/0" ]
  }
  tags {
    Name = "ssh-in"
  }
}
resource "aws_s3_bucket" "k8s-env" {
    bucket = "k8s-env"
    acl = "private"

    tags {
        Name = "k8s-env"
        Environment = "Dev"
    }
}
resource "aws_s3_bucket_object" "object" {
    bucket = "${aws_s3_bucket.k8s-env.bucket}"
    key = "${aws_instance.k8s-bootstrap.id}"
    source = "${var.env_file}"
    etag = "${md5(file("${var.env_file}"))}"
}
```
## Debug Output

https://gist.github.com/polasekr/239d987e061f11b8104360c04fc8d0e9
## Panic Output

n/a
## Expected Behavior

As shown in the output remote-exec provisioner is trying to connect to 172.16.1.84 (private IP address) via ssh while my expectation is that it should be using EC2 instance public IP address for this connection. I checked the documentation but I haven't found a way to control what IP address is used for the connection. This is crucial, since you are very likely building your AWS infrastructure from outside AWS cloud. I also checked the instance created by Terraform and it doesn't have public IP address yet. It appears that public IP allocation is happening after build completion.
### Actual Behavior

Instance build never gets completed.
### Steps to Reproduce

Please list the steps required to reproduce the issue, for example:
1. `terraform apply`

**Comment by apparentlymart** [comment]
Hi @polasekr!

[The `connection` block](https://www.terraform.io/docs/provisioners/connection.html) is the intended way to control how the provisioners connect to your instance(s).

You can use interpolation with the special `self` object to access the properties of the instance being provisioned, including the public IP address, like this:

```hcl
connection {
    host = "${self.public_ip}"
    // ... (and any other settings you need)
}
```

Terraform tries to provide some reasonable defaults, but in general I find that I almost always use a `connection` block for one reason or another. For example, authentication generally requires `connection` block attributes too.

**Comment by polasekr** [comment]
Hi Martin

Thank you for the prompt response.
I changed the code to include connection, but now I ran into dependency cycle. How should I define it to avoid dependency loop?

robert@razer-laptop:~/terraform_10point1$ terraform plan
Error configuring: 1 error(s) occurred:

* Cycle: aws_instance.k8s-bootstrap, aws_eip.k8s-bootstrap-IP

    provisioner "remote-exec" {
        inline = [
          "echo ${aws_instance.k8s-bootstrap.id} >${var.instancefile}",
          "apt-get install awscli s3cmd",
          "s3cmd --access_key=${var.access_key} --secret_key=${var.secret_key} --ssl get s3://k8s-env/${aws_instance.k8s-bootstrap.id}",
          "mkdir .ssh",
          "chmod 600 .ssh",
          "s3cmd --access_key=${var.access_key} --secret_key=${var.secret_key} --ssl get s3://k8s-env/${var.ssh_key} .ssh/id_rsa",
          "s3cmd --access_key=${var.access_key} --secret_key=${var.secret_key} --ssl -r get s3://k8s-env/${var.account_id} ~/.aws",
          "git clone git@github.com:10point1/k8s_build_script.git",
          "cd k8s_build_script",
          "chmod 755 k8s_env.sh",
          "./k8s_env.sh ~/${aws_instance.k8s-bootstrap.id} up >k8s_installation.log 2>&1"
        ]
        connection {
            private_key = "${file("/home/robert/.ssh/deployment.pem")}"
            host = "${aws_eip.k8s-bootstrap-IP.public_ip}"
        }
    }

**Comment by HX-Rd** [comment]
havent tried this my self but maybe try using the eip public_ip instead ${aws_eip.k8s-bootstrap-IP.public_ip} in the connection block. From the documentation "If you are using an aws_eip with your instance, you should refer to the EIP's address directly and not use public_ip, as this field will change after the EIP is attached."

**Comment by apparentlymart** [comment]
@polasekr sorry I didn't notice in your first message that you are using an `aws_eip` resource to assign an elastic IP to the instance.

You suspected correctly that the `aws_eip` resource won't be handled until the `aws_instance` resource is completed, and provisioning is part of creating the resource.

I think you have a few options here to work around this:
* Use `associate_public_ip_address` on your `aws_instance` to allocate it a regular (non-elastic IP) public IP address during provisioning, which I think will then change to the EIP once it's handled.
* Move your `provisioner` and `connection` blocks into the `aws_eip` resource rather than the `aws_instance` resource, so that the provisioning will happen only after the Elastic IP has attached, and will use the `public_ip` attribute of `aws_eip` instead of the EC2 instance.

The latter is probably the pattern we should describe in the docs, since I think it's the most robust workaround even though it feels a little strange to be "provisioning" an elastic IP address.

**Comment by polasekr** [comment]
@apparentlymart 

Thank you for your help. The solution of moving provisioner to aws_eip worked like a charm.

**Comment by apparentlymart** [comment]
I've re-tagged this as documentation to turn this ticket into a request to document that workaround. :)

Thanks!

**Comment by andrewrynhard** [comment]
@apparentlymart moving `provisioner` and `connection` to the `aws_eip` resource did not work for me, but moving them to `aws_eip_association` did.

**Comment by chris-brace** [comment]
In case any other newbs like myself end up here, this worked for me:

```resource "aws_instance" "bastion" {
  instance_type          = "t3.micro"
  ami                    = "${var.amis["ecs"]}"
  vpc_security_group_ids = ["${aws_security_group.bastion.id}"]
  key_name               = "us-east-2-root"
  subnet_id              = "${aws_subnet.public.*.id[0]}"

  tags = {
    Name = "${var.name}-bastion"
  }
}

resource "aws_eip" "bastion" {
  connection {
    user = "ec2-user"
    host = "${aws_eip.bastion.public_ip}"
  }

  instance = "${aws_instance.bastion.id}"

  tags = {
    Name = "${var.name}-bastion-public"
  }

  provisioner "remote-exec" {
    inline = ["sudo yum update -y"]
  }
}
```

i think what i didnt realize in the past 24 hours of reading about terraform is that a resource and refer to its own computed attributes in the way the connection -> host attr does.

**Comment by ukreddy-erwin** [comment]
Facing similar issue with azure now.
"${self.public_ip}" not working
Neither 
${azurerm_public_ip.<publicIPresource>.ip_address}

**Comment by mildwonkey** [comment]
I am going to close this issue due to inactivity and age.

If there is still a question, I recommend the [the community forum](https://discuss.hashicorp.com/c/terraform-core), where there are far more people available to help. If there is a bug or you would like to make a feature request, please open a new issue and fill out the template.
Thanks!
